src/kem/kyber768/sampling.rs

```diff
@@ -37,28 +36,35 @@ pub fn sample_from_uniform_distribution(
     Err(BadRejectionSamplingRandomnessError)
 }
 
-pub fn sample_from_binomial_distribution(
-    sampling_coins: usize,
+pub fn sample_from_binomial_distribution_with_2_coins(
     randomness: &[u8],
 ) -> KyberPolynomialRingElement {
-    assert_eq!(randomness.len(), sampling_coins * 64);
+    assert_eq!(randomness.len(), 2 * 64);
 
     let mut sampled: KyberPolynomialRingElement = KyberPolynomialRingElement::ZERO;
 
-    for i in 0..sampled.len() {
-        let mut coin_tosses: u8 = 0;
-        for j in 0..sampling_coins {
-            coin_tosses += randomness.nth_bit(2 * i * sampling_coins + j);
-        }
-        let coin_tosses_a: KyberFieldElement = coin_tosses.into();
+    for (chunk_number, byte_chunk) in randomness.chunks(4).enumerate() {
+        let random_bits_as_u32: u32 = (byte_chunk[0] as u32)
+            | (byte_chunk[1] as u32) << 8
+            | (byte_chunk[2] as u32) << 16
+            | (byte_chunk[3] as u32) << 24;
 
-        coin_tosses = 0;
-        for j in 0..sampling_coins {
-            coin_tosses += randomness.nth_bit(2 * i * sampling_coins + sampling_coins + j);
-        }
-        let coin_tosses_b: KyberFieldElement = coin_tosses.into();
+        let even_bits = random_bits_as_u32 & 0x55555555;
+        let odd_bits = (random_bits_as_u32 >> 1) & 0x55555555;
+
+        let coin_toss_outcomes = even_bits + odd_bits;
 
-        sampled[i] = coin_tosses_a - coin_tosses_b;
+        for outcome_set in (0..u32::BITS).step_by(4) {
+            let outcome_1: i16 = ((coin_toss_outcomes >> outcome_set) & 0x3)
+                .try_into()
+                .unwrap();
+            let outcome_2: i16 = ((coin_toss_outcomes >> (outcome_set + 2)) & 0x3)
+                .try_into()
```

Comment:
We could use `as i16` here. These things will need proofs for correctness anyway.

Reply:
Switched to using u16 with a new implementation of field element subtraction.